ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/KeccakHasher.java

```diff
@@ -14,19 +14,35 @@
  */
 package org.hyperledger.besu.ethereum.mainnet;
 
+import java.security.MessageDigest;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.bouncycastle.jcajce.provider.digest.Keccak;
+
+/** Hasher for Keccak-256 PoW. */
 public class KeccakHasher implements PoWHasher {
 
   public static final KeccakHasher KECCAK256 = new KeccakHasher();
 
   private KeccakHasher() {}
 
+  private static final ThreadLocal<MessageDigest> KECCAK_256 =
+      ThreadLocal.withInitial(Keccak.Digest256::new);
+
   @Override
   public void hash(
       final byte[] buffer,
       final long nonce,
       final long number,
       final EpochCalculator epochCalc,
       final byte[] headerHash) {
-    throw new UnsupportedOperationException("TODO"); // TODO
+    Bytes input = Bytes.wrap(Bytes.wrap(headerHash), Bytes.ofUnsignedLong(nonce));
+
+    MessageDigest digest = KECCAK_256.get();
+    digest.update(input.toArrayUnsafe());
```

Comment:
Where's the mixhash?  To replace the POW you replace the step where MixHash is sent to Dagger/hashimoto and instead send mixhash to keccack.  That's the Ethereum way. 

Doing it this way has deep implications to block header validation for the rest of the Besu client.  Preserving the meaning and utility of MixHash (which isn't part of Ethash, just what feeds the Ethash) is necessary for the stability of Besu.

Reply:
Hi folks, author of ECIP-1049 here just weighing in. The proposal has compatibility as its top priority, so there should be no changes to the block construction itself. Also the way the work is verified currently, using keccak256 of mixhash and nonce. The way I envision the system in the proposal is:

* Block is valid if `keccak256(keccak256(rlp(unsealed header)), nonce) <=  2^256 / difficulty`
* `mixhash = keccak256(rlp(unsealed header))` which is stored in the block. Currently this process is `ethash(rlp(unsealed header))`
*  So to validate a block we check `keccak256(mixhash, nonce) <= 2^256 / difficulty`

Let me know if this mental model is valid, since this is what I am basing my analysis of the code on.

=================================

Left 2 comments in the code as well.

Reply:
So the Implementation section of ECIP-1049 doesn't have that level of clarity (this one correct? https://github.com/ethereumclassic/ECIPs/issues/13).  It seems to be an extension of the rational section.  That is perhaps my biggest beef in that the spec doesn't in fact specify anything with clarity. It has some hints but is missing actual details, such as what is H sub 'n bar'?  

There needs to be a "specification" section in ECIP-1049, as noted by the [ecip template](https://github.com/ethereumclassic/ECIPs/blob/master/ecip-X.md).  These three bullet points should be in that section.  There also needs to be clear specification as to the proposed difficulty bump.  And some justification of the 100x difficulty bump, it appears to be pulled from thin air.

Also, there are so many versions of ECIP floating around I think the PR description and ultimate merge commit needs to reference the "official" location of the ECIP it's implmementing.

Reply:
I sincerely apologize about any confusion with the proposal, the roll out plan is this ECIP is officially here: https://github.com/ethereumclassic/ECIPs/issues/394. This is the re-draft that will be finalized into ECIP-1049, once we have a block number we feel comfortable to activate on. I'm adding to it based on your points, and welcome the feedback.

Once this PR to Besu is merged, and we have a number of hardware engineers, and other full node teams (Mantis, and core-geth) that'll base their implementation of nodes and hardware/mining pools software off of what Besu supports, and then we'll activate mainnet once it's supported by the nodes.

* The 100x difficulty bump is based on discussions with experienced chip designer and former ATI Technologies VP Henry Quan [report here](https://github.com/epicblockchain/whitepapers/blob/master/ETC/SHA3/SHA-3_Transition_Whitepaper_Impact_on_Ethereum_Classic_Mining_Hardware_and_Network_Security.md), and was selected on estimates of how much higher  we predict initial Keccak-256 hash rate will be compared to Ethash, when first launched.
* H sub n refers to nonce.